filebeat/input/filestream/input_integration_test.go

```diff
@@ -315,3 +315,137 @@ func TestFilestreamCloseTimeout(t *testing.T) {
+		"close.on_state_change.check_interval": "100ms",
+		"close.on_state_change.inactive":       "2s",
+	})
+
+	testlines := []byte("first line\nsecond line\nthird line\n")
+	env.mustWriteLinesToFile(testlogName, testlines)
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	env.startInput(ctx, inp)
+
+	env.waitUntilEventCount(3)
+	env.requireOffsetInRegistry(testlogName, len(testlines))
+	env.waitUntilHarvesterIsDone()
+
+	cancelInput()
+	env.waitUntilInputStops()
+}
+
+// test_close_inactive_file_removal from test_input.py
+func TestFilestreamCloseAfterIntervalRemoved(t *testing.T) {
+	env := newInputTestingEnvironment(t)
+
+	testlogName := "test.log"
+	inp := env.mustCreateInput(map[string]interface{}{
+		"paths":                                []string{env.abspath(testlogName)},
+		"prospector.scanner.check_interval":    "24h",
+		"close.on_state_change.check_interval": "10ms",
+		"close.on_state_change.inactive":       "100ms",
+		// reader is not stopped when file is removed to see if the reader can still detect
+		// if the file has been inactive even if it have been removed in the meantime
+		"close.on_state_change.removed": "false",
+	})
+
+	testlines := []byte("first line\nsecond line\nthird line\n")
+	env.mustWriteLinesToFile(testlogName, testlines)
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	env.startInput(ctx, inp)
+
+	env.waitUntilEventCount(3)
+	env.requireOffsetInRegistry(testlogName, len(testlines))
+
+	env.mustRemoveFile(testlogName)
+
+	env.waitUntilHarvesterIsDone()
+
+	cancelInput()
+	env.waitUntilInputStops()
+}
+
+// test_close_inactive_file_rotation_and_removal from test_input.py
+func TestFilestreamCloseAfterIntervalRotatedAndRemoved(t *testing.T) {
+	env := newInputTestingEnvironment(t)
+
+	testlogName := "test.log"
+	inp := env.mustCreateInput(map[string]interface{}{
+		"paths":                                []string{env.abspath(testlogName)},
+		"prospector.scanner.check_interval":    "24h",
+		"close.on_state_change.check_interval": "10ms",
+		"close.on_state_change.inactive":       "100ms",
+		// reader is not stopped when file is removed to see if the reader can still detect
+		// if the file has been inactive even if it have been removed in the meantime
+		"close.on_state_change.removed": "false",
+	})
+
+	testlines := []byte("first line\nsecond line\nthird line\n")
+	env.mustWriteLinesToFile(testlogName, testlines)
+
+	ctx, cancelInput := context.WithCancel(context.Background())
+	env.startInput(ctx, inp)
+
+	env.waitUntilEventCount(3)
+	env.requireOffsetInRegistry(testlogName, len(testlines))
+
+	newFileName := "test_rotated.log"
+	env.mustRenameFile(testlogName, newFileName)
+	env.mustRemoveFile(newFileName)
+
+	env.waitUntilHarvesterIsDone()
+
+	cancelInput()
+	env.waitUntilInputStops()
+}
+
+// test_close_inactive_file_rotation_and_removal2 from test_input.py
+func TestFilestreamCloseAfterIntervalRotatedAndNewRemoved(t *testing.T) {
```

Comment:
Was there a bug/issue that motivated this test in the python code base? It kinda looks redundant to what we already have, especially due to the file being updated in between the remove/renamed actions.

Reply:
The tests were added together in 2016-2017 and I assume they were inspired by these issues:
- https://github.com/elastic/filebeat/issues/271
- https://github.com/elastic/beats/issues/922

I am OK with removing the second one because it is indeed somewhat redundant.

Reply:
The first issue sounds like an inode reuse issue. As you heavily depend on the filesystem here, the outcome of the test is uncertain... depending on the host system this might be a quite flaky test, or a stable test. It is rather a case you want to simulate by mocking the file system, such that you can produce predictable file meta-data with and without inode reuse. This would be a great set of tests to have in general, but the integration test as is does not proof much. Plus this test does rename the file but the file is not deleted until the very end. In this test setup there is no chance to trigger an inode reuse.

The second issue is a mix of ignore_older and general performance issues it seem. 


With us having some well defined state and no chance of reuse between the harvester starts, I don't see how the test is related to reproduce any of the described issues.

Reasoning inductively I would assume we have one test for the file being removed and a separate test for the file being renamed... no other operation. If the tests work in isolation, any combination of rename, with a final remove should work. As we are testing and don't proof the TestFilestreamCloseAfterIntervalRotatedAndRemoved is still nice to have. But checking the tests here, it seems like we are missing a test with the file being renamed only.

Reply:
I have added the extra test.